frontend/model/contracts/group.js

```diff
@@ -774,6 +783,92 @@ sbp('chelonia/defineContract', {
         // }
       }
     },
+    'gi.contracts/group/addChatRoom': {
+      validate: objectOf({
+        chatRoomID: string,
+        attributes: objectOf(chatRoomAttributes)
+      }),
+      process ({ data, meta }, { state }) {
+        const { name, type, privacyLevel } = data.attributes
+        Vue.set(state.chatRooms, data.chatRoomID, {
+          creator: meta.username,
+          name,
+          type,
+          privacyLevel,
+          deletedDate: null,
+          users: []
+        })
+        if (!state.generalChatRoomId) {
+          Vue.set(state, 'generalChatRoomId', data.chatRoomID)
+        }
+      }
+    },
+    'gi.contracts/group/deleteChatRoom': {
+      validate: (data, { getters, meta }) => {
+        objectOf({ chatRoomID: string })(data)
+
+        if (getters.getChatRooms[data.chatRoomID].creator !== meta.username) {
+          throw new TypeError(L('Only the channel creator can delete channel.'))
+        }
+      },
+      process ({ data, meta }, { state }) {
+        Vue.delete(state.chatRooms[data.chatRoomID])
+      }
+    },
+    'gi.contracts/group/leaveChatRoom': {
+      validate: objectOf({
+        chatRoomID: string,
+        member: string,
+        leavingGroup: boolean // if kicker is exists, it means group leaving
+      }),
+      process ({ data, meta }, { state }) {
+        Vue.set(state.chatRooms[data.chatRoomID], 'users',
+          state.chatRooms[data.chatRoomID].users.filter(u => u !== data.member))
+      },
+      async sideEffect ({ meta, data }, { state }) {
+        const rootState = sbp('state/vuex/state')
+        if (meta.username === rootState.loggedIn.username && !sbp('okTurtles.data/get', 'JOINING_GROUP')) {
+          const sendingData = data.leavingGroup
+            ? { member: data.member }
+            : { member: data.member, username: meta.username }
+          await sbp('gi.actions/chatroom/leave', { contractID: data.chatRoomID, data: sendingData })
+        }
+      }
+    },
+    'gi.contracts/group/joinChatRoom': {
+      validate: objectMaybeOf({
+        username: string,
+        chatRoomID: string
+      }),
+      process ({ data, meta }, { state }) {
+        const username = data.username || meta.username
+        const { chatRoomID } = data
+        Vue.set(state.chatRooms[chatRoomID], 'users', state.chatRooms[chatRoomID].users.concat([username]))
+      },
+      async sideEffect ({ meta, data }, { state }) {
+        const rootState = sbp('state/vuex/state')
+        const username = data.username || meta.username
+        if (username === rootState.loggedIn.username) {
+          if (!sbp('okTurtles.data/get', 'JOINING_GROUP') || sbp('okTurtles.data/get', 'READY_TO_JOIN_CHATROOM')) {
```

Comment:
Two issues:

1. This `'READY_TO_JOIN_CHATROOM'` should be removed (per the other comment).
2. Why is there an `!sbp('okTurtles.data/get', 'JOINING_GROUP')` here? I'm pretty sure that should be removed. We should automatically sync the contract if we're the one who is logged in an are processing this message.

Reply:
1. We definitely need `READY_TO_JOIN_CHATROOM`. But just need to set this `false` at the end of this sideEffect function scope.
2. I don't think we should for this case. Users need to ignore join/leave chatroom actions while he's joining group because of the following two reasons;
  a) Users can only join `General` chatroom after joining a group
  b) Users can only join/leave chatroom thru group contract

Reply:
> 1. We definitely need `READY_TO_JOIN_CHATROOM`

Why? What will happen if we don't include it?

> I don't think we should for this case. Users need to ignore join/leave chatroom actions while he's joining group because of the following two reasons;
> a) Users can only join `General` chatroom after joining a group
> b) Users can only join/leave chatroom thru group contract

I'm not convinced. When a user accepts an invite from the `Join.vue` page, `'JOINING_GROUP'` is set to `true`. And it waits until the full contract is synced, and also sends a message to join the general chatroom.

In that case, we should sync `data.chatRoomID` below, yet this code is preventing that.

So I think this entire `if (!sbp('okTurtles.data/get', 'JOINING_GROUP') || sbp('okTurtles.data/get', 'READY_TO_JOIN_CHATROOM')) {` should be removed.

What happens if you remove it? Does it break anything?

Reply:
> Why? What will happen if we don't include it?

When joining a group, we need to skip `joinChatRoom` and `leaveChatRoom` of the group contract.
That's because we make another `joinChatRoom` action after we join the group.
So using only `JOINING_GROUP`, we can not identify if the `joinChatRoom` action is new or old.
`new` means a new action to join `General` chatroom
`old` means actions which are already made before joining the group by anyone(me or any others)

> What happens if you remove it? Does it break anything?

We need to have it. If we remove those conditions it breaks.
1. It will make notifications every time they join the group. That's because we make `gi.action/chatroom/join` action in the sideEffect of `joinChatRoom` of group contract.
2. It will make users to join all the chatrooms even though they don't need to join any chatroom except `General`.

Reply:
> It will make notifications every time they join the group. That's because we make gi.action/chatroom/join action in the sideEffect of joinChatRoom of group contract.

You mean it will post a message to the chatroom that they joined ...? That's supposed to happen, right?

> It will make users to join all the chatrooms even though they don't need to join any chatroom except General.

How?

Below, it says `await sbp('gi.actions/contract/syncAndWait', data.chatRoomID)` - so it should only join them to the chatroom that they requested to join.

Can you describe to me the steps that I need to take to reproduce what you're describing if these conditions are removed?

Reply:
Thanks for the call @Silver-IT - to summarize, here's a scenario to test after commenting out the condition along with the lines to set `'JOINING_CHATROOM'` and `'READY_TO_JOIN_CHATROOM'`:

1. Create a group
2. Join 2 users (u1 and u2)
3. For both users, check the chatroom, see that you are joined to the #general chatroom
4. For u2, leave the group via the Group Settings
5. Rejoin the group using the invite link
6. Check the chatroom

In my tests, there was an error upon rejoining, related to unexpected events being received... I will try to investigate this and post my findings.

Reply:
Thanks for your time to take the call too. I will write detailed scenario in the next comment.
But first things first, you should understand the workflow.
1. User joins a chatroom thru `gi.contracts/group/joinChatRoom` and leaves a chatroom thru `gi.contracts/group/leaveChatRoom`. And every time user joins the group, we make a new action `gi.contracts/group/joinChatRoom` to join `General` chatroom. From this, we can say that users don't need to run the previous `joinChatRoom` and `leaveChatRoom` actions when they join a group.
2. We call the new `gi.contracts/group/joinChatRoom` action again to join `General` chatroom after we join a group. So we need to identify that action.

Reply:
Just summarizing the call, this code seems to work:

```js
        if (username === rootState.loggedIn.username) {
          // while users are joining chatroom, they don't need to leave chatrooms
          // this is similar to setting 'JOINING_GROUP' before joining group
          sbp('okTurtles.data/set', 'JOINING_CHATROOM', true)
          await sbp('gi.actions/contract/syncAndWait', data.chatRoomID)
          sbp('okTurtles.data/set', 'JOINING_CHATROOM', false)
        }
```

Reply:
I think you missed one thing.
As each user joins all the chatrooms (which they have ever joined before) while they joins the group, they sync all the chatroom contracts. This is unnecessary. :stuck_out_tongue:

Reply:
Hmm, tomorrow I will have a closer look at this.

I think this is because there are other guards elsewhere, for example, in `sideEffect` of `'gi.contracts/group/leaveChatRoom'`, there's:  `&& !sbp('okTurtles.data/get', 'JOINING_GROUP')`, and in `'gi.contracts/chatroom/leave'` there's `sbp('okTurtles.data/get', 'JOINING_CHATROOM'))`.

Maybe if these are also removed this code will work? Or will that also cause other problems? (I will try testing tomorrow but maybe you can also check)

Reply:
OK, you're right. So, I also tested removing all of the `'JOINING_CHATROOM'` stuff from everywhere, and that doesn't work because:

The reason is because when we rejoin the group and the chatroom, and sync the chatroom contract for the chatroom that we left, we process `gi.contracts/chatroom/leave/sideEffect`, which removes the contract in the middle of the sync we're performing.

Then we get the `[CRITICAL ERROR] NOT EXPECTING EVENT!` error.

I also tested leaving `'JOINING_CHATROOM'` while commenting out the `if (!sbp('okTurtles.data/get', 'JOINING_GROUP') || sbp('okTurtles.data/get', 'READY_TO_JOIN_CHATROOM'))` condition and `READY_TO_JOIN_CHATROOM` stuff, and you're correct also that we will in that case, upon rejoining the group, sync chatrooms that we're not a part of anymore. So OK, this good. I'm considering this thread "resolved" but leaving the thread open so that we don't lose the discussion, as I think this is important to remember for me and for others to understand this code.

Reply:
👍